### PR TITLE
Add all files

### PR DESCRIPTION
**Pull request by JasonHoffmann**

### CODE REVIEW

File: geovin/includes/class-add-product-type.php [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+        add_action( 'woocommerce_after_variations_table', array( $this, 'add_cart_image_input' ) );
+
+        add_filter( 'woocommerce_add_cart_item_data', array( $this, 'add_cart_image_data' ), 10, 4 );
+
+        add_filter( 'woocommerce_cart_item_thumbnail', array( $this, 'adjust_cart_thumb' ), 10, 3 );
+
+    }
+
+    public function adjust_cart_thumb( $image, $cart_item, $cart_item_key ) {
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:48]: This is a pattern that I've noticed throughout, likely because it is just you maintaining this code, but code comments can help me understand the way each of these are used, and what to expect from arguments passed to it.

File: geovin/includes/class-add-product-type.php [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        if ( isset( $cart_item['image_to_use'] ) && $cart_item['image_to_use'] !== '' ) {
+            $image = '<img width="300" height="300" src="' . $cart_item['image_to_use'] . '" class="attachment-woocommerce_thumbnail size-woocommerce_thumbnail" alt="" loading="lazy" />';
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:50]: Any code output to the front-end should be escaped, i.e. `esc_attr`

File: geovin/includes/class-add-product-type.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        }
+        return $image;
+    }
+
+    public function add_cart_image_data( $cart_item_data, $product_id, $variation_id, $quantity ) {
+        $cart_item_data['image_to_use'] = $_REQUEST['cart_image_to_use'];
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:56]: When processing responses from the `$_REQUEST` object, data should be sanitized, for instance with `sanitize_text_field` or through the use of `filter_input`.

Also, it is helpful to use either `$_POST` or `$_GET`  (or `INPUT_GET` or `INPUT_POST`) rather than a full request object to prevent data from being spoofed.

File: geovin/includes/class-add-product-type.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        ob_start();
+
+        if ( empty( $_POST['sku'] ) ) {
+            wp_die();
+        }
+        $variable_product_id = wc_get_product_id_by_sku( $_POST['sku'] );
+        $variable_product = wc_get_product( absint( $_POST['product_id'] ) );
+
+        if ( ! $variable_product ) {
+            wp_die();
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:181]: This could likely benefit from better error handling, and potentially passing feedback to the admin or to the user about what went wrong.

File: geovin/includes/class-add-product-type.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        if ( ! get_term_by( 'slug', 'geovin', 'product_type' ) ) {
+            wp_insert_term( 'geovin', 'product_type' );
+        }
+    }
+
+    /**
+     * Get a matching variation based on sku.
+     */
+    public static function get_variation_from_sku() {
+        ob_start();
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:172]: Beginning output buffering without ending it can cause some unintended side effects.

File: geovin/includes/class-add-product-type.php [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /**
+     * Installing on activation
+     *
+     * @return void
+     */
+    public function install() {
+        // If there is no advanced product type taxonomy, add it.
+        if ( ! get_term_by( 'slug', 'geovin', 'product_type' ) ) {
```
Comment (JasonHoffmann) [inline geovin/includes/class-add-product-type.php:163]: If this term is (accidentally) deleted, but the activation hook isn't called again through a deactivation and reactivation, what happens? Can this continue to work without this specific term in there?

File: geovin/includes/class-cli.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+			}
+		}
+
+	    $uploads  = wp_upload_dir( null, false );
+		$logs_dir = $uploads['basedir'] . '/product-exports/' . date('mY');
+		if ( ! is_dir( $logs_dir ) ) {
+		    mkdir( $logs_dir, 0755, true );
+		}
+	    $file_resource = fopen( $logs_dir . '/trending-products.csv', 'a' );
+		WP_CLI\Utils\write_csv( $file_resource, $rows, array('name', 'code', 'link') );
```
Comment (JasonHoffmann) [inline geovin/includes/class-cli.php:98]: This is a handy utility!

File: geovin/includes/class-cart.php [hunk 8]
```diff
@@ -0,0 +1,1031 @@
...
+        $term_id = $term->term_id;
+        $term_id_prefixed = $taxonomy_prefix .'_'. $term_id;
+        $geovin_code = get_field( 'code', $term_id_prefixed );
+
+        return $geovin_code;
+	}
+
+	public function add_copies( $recipient, $order, $email ) {
+		$send_to = $order->get_meta('_billing_sales_order_send_to');
+		if ( $send_to ) {
```
Comment (JasonHoffmann) [inline geovin/includes/class-cart.php:485]: An `! empty` check can account for situations where there is a value (for instance a `0` or `true`) but you still don't want to add it.

File: geovin/includes/class-cart.php [hunk 9]
```diff
@@ -0,0 +1,1031 @@
...
+			$paramlocalfields_array->postcode = str_replace('#billing_postcode_field, ', '', $paramlocalfields_array->postcode);
+			$paramlocalfields_array->state = str_replace('#billing_state_field, ', '', $paramlocalfields_array->state);
+			$params['locale_fields'] = json_encode($paramlocalfields_array);
+		}
+		
+
+		return $params;
+	}
+
+	public function add_address_dropdown( $checkout ) {
```
Comment (JasonHoffmann) [inline geovin/includes/class-cart.php:574]: This seems like a really cool extension to WooCommerce, and this is a good use of static functions to retrieve the data. It makes me think if it would be possible to provide this same functionality via a custom REST endpoint and then build things on the front-end via JavaScript and React, which might give you a bit more control over the select.

File: geovin/sd-v3/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,1167 @@
...
+        this.initial_parameter_values = values;
+        $('.sd-curtain span').text('Initializing your model...');
+        $('.loading-bar__inner').css({width:'33%'});
+        $('.loading-bar__inner').text('33%');
+        
+        await this.createViewport();
+        await this.createSession( this.sd_ticket );
+        this.initialized = true;
+        //await this.addDefaultScene();
+        this.availableScenes = this.currentViewport.lightScenes;
```
Comment (JasonHoffmann) [inline geovin/sd-v3/src/index.ts:844]: This may be a good opportunity to do conditional checks on `this.currentViewport` (i.e. `this.currentViewport?.lightScenes`), since the currentViewport is being loaded in via an async function.

File: geovin/sd-v3/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,1167 @@
...
+        await this.currentViewport.render();
+    }
+
+    private async enableZoom() {
+        const camera = this.currentViewport.camera!;
+        camera.enableZoom = true;
+        await this.currentViewport.render();
+    }
+
+    private async adjustCamera() {
```
Comment (JasonHoffmann) [inline geovin/sd-v3/src/index.ts:960]: I like the way these functions are broken down, taking advantage of multiple threads being able to fire off at once.

File: geovin/sd-v3/build/sd-v3/src/index.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+            //console.log('light scenes',this.availableScenes,this.activeScene);
+            $('.sd-curtain span').text('Applying selections ...');
+            $('.loading-bar__inner').css({ width: '66%' });
+            $('.loading-bar__inner').text('66%');
+            const task_token = yield SDV.addListener(SDV.EVENTTYPE.TASK.TASK_END, (task) => {
+                //console.log('task token',task_token,task );
+                if (task.type === 'session_customization') {
+                    const rendering_token = SDV.addListener(SDV.EVENTTYPE.RENDERING.BEAUTY_RENDERING_FINISHED, (render) => {
+                        //console.log('event token',rendering_token,render );
+                        this.rendering = false;
```
Comment (JasonHoffmann) [inline geovin/sd-v3/build/sd-v3/src/index.js:56]: If the function never makes it here, does the constructor get stuck on `this.rendering`? Is there a way to clear that elsewhere if all else fails?

File: geovin/sd-v3/build/sd-v3/src/index.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            $('.loading-bar__inner').css({ width: '33%' });
+            $('.loading-bar__inner').text('33%');
+            yield this.createViewport();
+            yield this.createSession(this.sd_ticket);
+            this.initialized = true;
+            this.availableScenes = this.currentViewport.lightScenes;
+            this.activeScene = this.currentViewport.lightScene;
+            //console.log('light scenes',this.availableScenes,this.activeScene);
+            $('.sd-curtain span').text('Applying selections ...');
+            $('.loading-bar__inner').css({ width: '66%' });
```
Comment (JasonHoffmann) [inline geovin/sd-v3/build/sd-v3/src/index.js:49]: Anyway to get real time feedback on the percentages here? Or to present the loading state in a different way?

I don't have all the info on how long this typically takes but there could be another approach here. The text feels more helpful because it provides step by step directions.

**Review by JasonHoffmann — COMMENTED** [review]
Overall, this code is very through and accounts for many different possible edge cases. It has a solid integration into WooCommerce and is ready (from my perspective) for everyday production use. I particularly like that it extends things in a way that feels very native to WooCommerce and so fits right in alongside other plugins.

There are a few things I would point to as general areas of improvement:
1. The code in general could benefit from code documentation, outlining the uses and intentions behind functionality
2. Most of the functionality seems to follow the happy path, but doesn't always account for error states. That being said there are a few crucial places where edge cases and errors are accounted for and those make it more resilient
3. Sanitization and escaping functions aren't always consistently used.

I've also found a handful of files and added comments to specific lines to point out some examples of the above, plus a few other things that I noticed.

File: geovin/includes/class-add-product-type.php (hunk 4 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:181]: Agreed - when I went to investigate this one and why I didn't use `wp_send_json_success()` or `wp_send_json_error()` I realized it's a match to a similar function Woo has for variable products: [get_variation()](https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-ajax.html#source-view.550). 

Geovin's attribute options directly relate to a SKU for each possible configuration, so it's much easier and faster to programmatically construct a SKU and then look up the product based on the SKU instead of Woo's default approach which essentially runs a taxonomy query on the db to find the product that has all of the attributes the user selected. Geovin has some products with over 20,000 variations... so all that to say this is a swap out function for better performance.

Given that, I'm inclined to leave it as is so that it stays consistent with Woo. Woo does output text to the user if no variation is returned. I'm curious what your take is - would you typically leave it to match the plugin your extending, or go ahead and make the improvements to clarify success/fail?

File: geovin/includes/class-add-product-type.php (hunk 5 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:172]: I thought this was perhaps something I did inadvertently changing my mind on how I was going deliver this data for the ajax, but I was really surprised to see this is also in the original Woo function I swapped out ([see my comment here](https://github.com/reaktivstudios/geovin/pull/1#discussion_r2258593363)) - but this is a scenario where I think veering from Woo makes sense - I'm removing it.

File: geovin/includes/class-add-product-type.php (hunk 6 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:163]: That is an interesting question. There isn't a UI in the dashboard that allows managing of the product_type terms, so I hadn't considered this possibility. I did some digging, and found two things: 

1. If the term did get deleted, it seems as though things would still function as expected and Woo would add the term the next time a product of that type was saved (the type would still appear in the product type dropdown selector as that isn't populated by the terms directly) I definitely intend to test this scenario now out of just sheer curiosity!
2. I also investigated how Woo is adding their default product_type terms - they are added on install, and it looks as though they run the install function again anytime an update is needed. I do think I can hook into the `woocommerce_installed` action to run my function at that point also so it always runs when the Woo create_terms function runs, so I'm going to add that.

File: geovin/includes/class-cli.php (hunk 7 above)
Comment (krissie-northux) [reply geovin/includes/class-cli.php:98]: Yes! It's so helpful!!

File: geovin/includes/class-cart.php (hunk 9 above)
Comment (krissie-northux) [reply geovin/includes/class-cart.php:574]: Yes, I absolutely agree - this stretched their scope a bit to include and was added towards the end of the build, but it significantly improved the usability for their Dealers. I remember thinking it would be a "fun" project to add this in a way it could be more broadly distributed as a stand alone Woo plugin. I think a REST based implementation would be ideal, and would hopefully make it easy to include with Woo's checkout blocks.

File: geovin/includes/class-add-product-type.php (hunk 1 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:48]: Yes, definitely guilty of not adding comments when it's just me. I've gone through all of the code and added comments in this commit: https://github.com/krissie-northux/geovin/commit/37cf9c24dbc5c3561eb8b4178933f8766b6e0fb1

File: geovin/includes/class-add-product-type.php (hunk 2 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:50]: I combed through to find other instances where I had missed escaping variables being used for the front end and added adjusted them all in this commit: https://github.com/krissie-northux/geovin/commit/5e2b4cb89136d98cfafb0beb19f63c61d2eb1625

File: geovin/includes/class-add-product-type.php (hunk 3 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:56]: I had not made use of filter_input before, but went through and updated any place I had used the `$_REQUEST` object to use filter_input instead, those changes are in this commit: https://github.com/krissie-northux/geovin/commit/3c82aff127dfac01d0e666abe5ba1cb6e715c7d8

File: geovin/includes/class-add-product-type.php (hunk 5 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:172]: removed it in this commit: https://github.com/krissie-northux/geovin/commit/c6bcef485744ce523df400f0e5b2175d96cfea55

File: geovin/includes/class-cart.php (hunk 8 above)
Comment (krissie-northux) [reply geovin/includes/class-cart.php:485]: there were several places in this file where using `! empty` was a much better conditional check, I updated all of them in this commit: https://github.com/krissie-northux/geovin/commit/c1cc199f7589a9c98e34f9b0cf3d1b94ccc1e0bf

File: geovin/includes/class-add-product-type.php (hunk 6 above)
Comment (krissie-northux) [reply geovin/includes/class-add-product-type.php:163]: Here is the commit where I added my install function to the `woocommerce_installed` action: https://github.com/krissie-northux/geovin/commit/76c51ecf23ce4563d14d94fb080004e0a0b79d91

File: geovin/sd-v3/src/index.ts (hunk 10 above)
Comment (krissie-northux) [reply geovin/sd-v3/src/index.ts:844]: I believe because createViewport() is awaited directly before this line, and it assigns this.currentViewport after awaiting a Promise from SDV.createViewport(), that it will successfully return a valid object, and this.currentViewport should be reliably set at this point. My understanding is that await is kind of a short hand way of doing .then() on a function that returns a Promise. Thoughts?

The use of `.?` as a conditional check is new to me, but it is a really nice and clean way to do it, and I read up a bit on the use of optional chaining - I will be utilizing this often.

File: geovin/sd-v3/src/index.ts (hunk 11 above)
Comment (krissie-northux) [reply geovin/sd-v3/src/index.ts:960]: Thanks, this benefited from essentially being a re-write of the first implementation, when the client wanted to upgrade to the Shapediver V3 Viewer - enough changed in the V3 api that it was a full re-write. It was nice to take advantage of the knowledge and experience of what worked well and didn't work well with the first implementation.

File: geovin/sd-v3/build/sd-v3/src/index.js (hunk 12 above)
Comment (krissie-northux) [reply geovin/sd-v3/build/sd-v3/src/index.js:56]: It would get stuck, but I kind of need it to in this instance. I use this boolean to verify if it's safe to close the viewer modal when a user finishes adjusting their specs. When the modal is closed, I adjust the camera view and take a screenshot for use on the product page and in the cart if they add that configuration to it. I can't take the screenshot if the viewer is still rendering the changes on the model, it will fail. So when `this.rendering` is `true` the close button is unavailable and an hourglass processing indicator appears, once Shapediver is finished it fires this event, and I can safely let them close the modal.

File: geovin/sd-v3/build/sd-v3/src/index.js (hunk 13 above)
Comment (krissie-northux) [reply geovin/sd-v3/build/sd-v3/src/index.js:49]: There wasn't at the time, but Shapediver may have added some additional events in the api I could tap into, they tend to add new features consistently. I will definitely look when I work on the project the client has coming - they are adding a new furniture line, so I'll need to be adjusting this file specifically. The Shapediver plan that the client is on, put them at the bottom of the priority list related to server resources in rendering their models. During peek times the rendering would sometimes take just long enough for users to question if anything was happening. This was a way to at least let the user know something was going on, despite it not being a true loading progress indicator - we just needed to buy 2-5 seconds of the users attention for those peek times.

File: geovin/includes/class-add-product-type.php (hunk 1 above)
Comment (JasonHoffmann) [reply geovin/includes/class-add-product-type.php:48]: These comments are great, and make it that much easier to follow, thank youl.

File: geovin/includes/class-add-product-type.php (hunk 4 above)
Comment (JasonHoffmann) [reply geovin/includes/class-add-product-type.php:181]: I think one approach would be to store instances when a product is not found to the database in some way in a log so that you can follow patterns of it happening. So maybe not adjusting the function overall to keep it consistent, but providing yourself a way to know how often it's occuring and if further adjustments would be needed.

File: geovin/includes/class-add-product-type.php (hunk 6 above)
Comment (JasonHoffmann) [reply geovin/includes/class-add-product-type.php:163]: Makes sense!

File: geovin/sd-v3/build/sd-v3/src/index.js (hunk 12 above)
Comment (JasonHoffmann) [reply geovin/sd-v3/build/sd-v3/src/index.js:56]: Got it, ok that follows

File: geovin/sd-v3/build/sd-v3/src/index.js (hunk 13 above)
Comment (JasonHoffmann) [reply geovin/sd-v3/build/sd-v3/src/index.js:49]: Fully understood, feedback like this is certainly better than no feedback at all.
